packages/protocol/contracts/governance/DowntimeSlasher.sol

```diff
@@ -35,88 +42,212 @@ contract DowntimeSlasher is SlasherUtil {
     );
-    require(endSignerIndex < numberValidatorsInSet(endBlock), "Bad validator index at end block");
-    address startSigner = validatorSignerAddressFromSet(startSignerIndex, startBlock);
-    address endSigner = validatorSignerAddressFromSet(endSignerIndex, endBlock);
-    IAccounts accounts = getAccounts();
+
+    return (bitmaps[msg.sender][startBlock][endBlock] & bytes32(1 << signerIndex)) == 0;
+  }
+
+  /**
+   * @notice Shows if the user already called setBitmapForInterval for
+   * the specific interval.
+   * @param startBlock First block of a calculated downtime interval.
+   * @param endBlock Last block of the calculated downtime interval.
+   * @return True if the user already called setBitmapForInterval for
+   * the specific interval.
+   */
+  function isBitmapSetForInterval(uint256 startBlock, uint256 endBlock) public view returns (bool) {
+    // It's impossible to have all the validators down in a interval.
+    return bitmaps[msg.sender][startBlock][endBlock] != 0;
+  }
+
+  /**
+   * @notice Returns true if a validator has been down for the specified overlapping or adjacent
+   * intervals.
+   * @param startBlocks startBlocks of the specified intervals.
+   * @param endBlocks endBlocks of the specified intervals.
+   * @param signerIndices Indices of the signers within the validator set for every epoch change.
+   * @return True if the validator signature does not appear in any block within the window.
+   */
+  function wasDownForIntervals(
+    uint256[] memory startBlocks,
+    uint256[] memory endBlocks,
+    uint256[] memory signerIndices
+  ) public view returns (bool) {
+    require(startBlocks.length > 0, "requires at least one interval");
     require(
-      accounts.signerToAccount(startSigner) == accounts.signerToAccount(endSigner),
-      "Signers do not match"
+      startBlocks.length == endBlocks.length,
+      "startBlocks and endBlocks must have the same length"
     );
-    uint256 sz = getEpochSize();
-    uint256 startEpoch = epochNumberOfBlock(startBlock, sz);
-    for (uint256 n = startBlock; n <= endBlock; n = n.add(1)) {
-      uint256 signerIndex = epochNumberOfBlock(n, sz) == startEpoch
-        ? startSignerIndex
-        : endSignerIndex;
-      // We want to check signers for block n,
-      // so we get the parent seal bitmap for the next block
-      if (uint256(getParentSealBitmap(n.add(1))) & (1 << signerIndex) != 0) return false;
+    require(signerIndices.length > 0, "requires at least one signerIndex");
+
+    uint256 epochSize = getEpochSize();
+    uint256 signerIndicesIndex = 0;
+    for (uint256 i = 0; i < startBlocks.length; i = i.add(1)) {
+      if (i > 0) {
+        require(
+          startBlocks[i.sub(1)] < startBlocks[i],
+          "each interval must start after the previous interval's start"
+        );
+        require(
+          startBlocks[i] <= endBlocks[i.sub(1)].add(1),
+          "each interval must start at least one block after previous interval's end"
```

Comment:
nit: `each interval must start at most one block after the end of the previous interval`